src/rust/catloop/mod.rs

```diff
@@ -0,0 +1,521 @@
+            Some(None) => {
+                self.qtable.free(qd);
+                Ok(())
+            },
+
+            // Socket is bound to a duplex pipe.
+            Some(Some((_, duplex_pipe))) => {
+                // Check if this is an active socket.
+                if !self.listening.remove(&qd) {
+                    // Push magic close sequence.
+                    let sga: demi_sgarray_t = self.cook_magic_close()?;
+                    let qt: QToken = duplex_pipe.push(&sga)?;
+                    self.catmem.borrow_mut().free_sgarray(sga)?;
+
+                    // Wait until magic sequence is pushed.
+                    while DuplexPipe::poll(&self.catmem, qt)?.is_none() {
+                        // no operation
+                    }
+                }
+
+                // Close duplex pipe.
+                duplex_pipe.close()?;
+                self.qtable.free(qd);
+                Ok(())
+            },
+            None => Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
+        }
+    }
+
+    /// Pushes a scatter-gather array to a socket.
+    pub fn push(&mut self, qd: QDesc, sga: &demi_sgarray_t) -> Result<QToken, Fail> {
+        trace!("push() qd={:?}", qd);
+
+        let catmem_qd: QDesc = match self.sockets.get(&qd) {
+            Some(&Some((_, ref duplex_pipe))) => duplex_pipe.tx(),
+            Some(&None) => return Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
+            None => qd,
+        };
+
+        let qt: QToken = self.catmem.borrow_mut().push(catmem_qd, sga)?;
+        self.qts.insert(qt, (demi_opcode_t::DEMI_OPC_PUSH, qd));
+
+        Ok(qt)
+    }
+
+    /// Pops data from a socket.
+    pub fn pop(&mut self, qd: QDesc) -> Result<QToken, Fail> {
+        trace!("pop() qd={:?}", qd);
+
+        let catmem_qd: QDesc = match self.sockets.get(&qd) {
+            Some(&Some((_, ref duplex_pipe))) => duplex_pipe.rx(),
+            Some(&None) => return Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
+            None => qd,
+        };
+
+        let qt: QToken = self.catmem.borrow_mut().pop(catmem_qd)?;
+        self.qts.insert(qt, (demi_opcode_t::DEMI_OPC_POP, qd));
+
+        Ok(qt)
+    }
+
+    /// Allocates a scatter-gather array.
+    pub fn sgaalloc(&self, size: usize) -> Result<demi_sgarray_t, Fail> {
+        self.catmem.borrow_mut().alloc_sgarray(size)
+    }
+
+    /// Releases a scatter-gather array.
+    pub fn sgafree(&self, mut sga: demi_sgarray_t) -> Result<(), Fail> {
+        // Sanity check length of scatter-gather array.
+        // ToDo: Drop this check once we support scatter-gather arrays with multiple segments.
+        if sga.sga_numsegs != 1 {
+            return Err(Fail::new(libc::EINVAL, "demi_sgarray_t has invalid segment count"));
+        }
+
+        // Check for magic close message.
+        if sga.sga_segs[0].sgaseg_len == 0 {
+            sga.sga_segs[0].sgaseg_len = mem::size_of_val(&Self::MAGIC_CLOSE) as u32;
```

Comment:
This looks dangerous.  Can there be a legitimate zero-length scatter-gather array here?  If so, this could lead to a segmentation fault in `is_magic_close` below.

Reply:
No, we cannot have a legit zero-length scatter-gather array thus segmentation fault is not possible.